firedrake/functionspace.py

```diff
@@ -732,26 +757,48 @@ def __init__(self, spaces, name=None):
             ME  = MixedFunctionSpace([P2v, P1, P1, P1])
         """
 
-        if self._initialized:
-            return
-        self._spaces = [IndexedFunctionSpace(s, i, self)
-                        for i, s in enumerate(flatten(spaces))]
-        self._mesh = self._spaces[0].mesh()
-        self._ufl_element = ufl.MixedElement(*[fs.ufl_element() for fs in self._spaces])
-        self.name = name or '_'.join(str(s.name) for s in self._spaces)
-        self.rank = 1
-        self._index = None
-        self._initialized = True
-        dm = PETSc.DMShell().create()
-        from firedrake.function import Function
-        with Function(self).dat.vec_ro as v:
-            dm.setGlobalVector(v.duplicate())
-        dm.setAttr('__fs__', weakref.ref(self))
-        dm.setCreateFieldDecomposition(self.create_field_decomp)
-        dm.setCreateSubDM(self.create_subdm)
-        self._dm = dm
-        self._ises = self.dof_dset.field_ises
-        self._subspaces = []
+        # Check that function spaces are on the same mesh
+        meshes = [space.mesh() for space in spaces]
+        for i in xrange(1, len(meshes)):
+            if meshes[i] is not meshes[0]:
+                raise ValueError("All function spaces must be defined on the same mesh!")
```

Comment:
Currently every function space class is topological, and `WithGeometry` adds a mesh geometry. Therefore `MixedFunctionSpace` is topological, so it has topological function spaces as subspaces. A `WithGeometry` can surround a `MixedFunctionSpace` to add a mesh geometry.

I didn't know Firedrake can solve problems on different meshes. Can we add a test that exercises the feature that the Leeds people use?